### PR TITLE
Fix Issue 18215 - std.array.replace throws a range violation if the from range is longer than the array

### PR DESCRIPTION
**Pull request by wilzbach**
Maybe I'm just too tired, but how is it possible that there wasn't a test for replacing on an array of arrays?

### CODE REVIEW

File: std/array.d [hunk 1]
```diff
@@ -2146,13 +2151,26 @@ if (isDynamicArray!(E[]) && isForwardRange!R1 && isForwardRange!R2
...
+@safe unittest
+{
+    auto arr = ["aaa.dd", "b"];
+    arr = arr.replace("aaa.dd", ".");
+    assert(arr == [".", "b"]);
+
+    arr = ["_", "_", "aaa.dd", "b", "c", "aaa.dd", "e"];
+    arr = arr.replace("aaa.dd", ".");
+    assert(arr == ["_", "_", ".", "b", "c", ".", "e"]);
+}
```
Comment (quickfur) [inline std/array.d]: Is there already a unittest that covers the case `"abababc".replace("abcdef","g")`? I.e., the second branch of the static if.

Comment (wilzbach) [reply std/array.d]: Yeah that's the original case and covered by the existing tests. Expand the diff above.

image

BTW you might want to install the CodeCov browser extension: https://github.com/codecov/browser-extension

Comment (wilzbach) [reply std/array.d]: I still added more tests ;-)

Comment (quickfur) [reply std/array.d]: Yeah I was confused by your change to use `min(from.length, $)`, which you just reverted. I suppose that was not necessary? :-P

File: std/array.d [hunk 2]
```diff
@@ -2146,13 +2151,42 @@ if (isDynamicArray!(E[]) && isForwardRange!R1 && isForwardRange!R2
...
+
+// https://issues.dlang.org/show_bug.cgi?id=18215
+@safe unittest
+{
+    assert([[0], [1, 2], [0], [3]].replace([0], [4]) == [[4], [1, 2], [4], [3]]);
+    assert([[0], [1, 2], [0], [3], [1, 2]].replace([1, 2], [0]) == [[0], [0], [0], [3], [0]]);
```
Comment (quickfur) [inline std/array.d]: Nitpick: This line is a little too long.

**Comment by wilzbach** [comment]
I just found https://issues.dlang.org/show_bug.cgi?id=10930 in Bugzilla:
It's about:

```d
[1, 2, 3].replace(2, 4);
```

With this fix, it will be easy to fix the issue as follows:

```d
diff --git a/std/array.d b/std/array.d
index 3a44612cb..65e161504 100644
--- a/std/array.d
+++ b/std/array.d
@@ -2120,21 +2120,28 @@ if (isInputRange!RoR &&
         $(REF map, std,algorithm,iteration) which can act as a lazy replace
  +/
 E[] replace(E, R1, R2)(E[] subject, R1 from, R2 to)
-if (isDynamicArray!(E[]) && isForwardRange!R1 && isForwardRange!R2
-        && (hasLength!R2 || isSomeString!R2))
+if (isDynamicArray!(E[]))
 {
     import std.algorithm.searching : find;
     import std.range : dropOne;
 
-    if (from.empty) return subject;
+    static if (isInputRange!R1)
+    {
+        if (from.empty) return subject;
+        alias rSave = a => a.save;
+    }
+    else
+    {
+        alias rSave = a => a;
+    }
 
-    auto balance = find(subject, from.save);
+    auto balance = find(subject, rSave(from));
     if (balance.empty)
         return subject;
 
     auto app = appender!(E[])();
     app.put(subject[0 .. subject.length - balance.length]);
-    app.put(to.save);
+    app.put(rSave(to));
     // replacing an element in an array is different to a range replacement
     static if (is(Unqual!E : Unqual!R1))
         replaceInto(app, balance.dropOne, from, to);
@@ -2158,6 +2165,11 @@ if (isDynamicArray!(E[]) && isForwardRange!R1 && isForwardRange!R2
     assert([3, 3, 4, 3].replace([3, 3], [1, 1, 1]) == [1, 1, 1, 4, 3]);
 }
 
+@safe unittest
+{
+    assert([0, 1, 2].replace(1, 4) == [0, 4, 2]);
+}
+
 // https://issues.dlang.org/show_bug.cgi?id=18215
 @safe unittest
 {
@@ -2181,28 +2193,34 @@ if (isDynamicArray!(E[]) && isForwardRange!R1 && isForwardRange!R2
 
 /// ditto
 void replaceInto(E, Sink, R1, R2)(Sink sink, E[] subject, R1 from, R2 to)
-if (isOutputRange!(Sink, E) && isDynamicArray!(E[])
-    && isForwardRange!R1 && isForwardRange!R2
-    && (hasLength!R2 || isSomeString!R2))
+if (isOutputRange!(Sink, E) && isDynamicArray!(E[]))
 {
     import std.algorithm.searching : find;
     import std.range : dropOne;
 
-    if (from.empty)
+    static if (isInputRange!R1)
+    {
+        if (from.empty)
+        {
+            sink.put(subject);
+            return;
+        }
+        alias rSave = a => a.save;
+    }
+    else
     {
-        sink.put(subject);
-        return;
+        alias rSave = a => a;
     }
     for (;;)
     {
-        auto balance = find(subject, from.save);
+        auto balance = find(subject, rSave(from));
         if (balance.empty)
         {
             sink.put(subject);
             break;
         }
         sink.put(subject[0 .. subject.length - balance.length]);
-        sink.put(to.save);
+        sink.put(rSave(to));
         // replacing an element in an array is different to a range replacement
         static if (is(Unqual!E : Unqual!R1))
             subject = balance.dropOne;
```

Of course constraints need be set, but to be fair currently there aren't any for this (and probably `save` should be called if `R1` is a range
I would do this in another PR for `master` as this is enhancement request (the `replace` currently rejects this), whereas `replace` currently accepts RoR (ranges of ranges), but returns invalid output.
I'm aware that `[1, 2, 3].replace(2.only, 4.only)` or `[1, 2, 3].replace([2], [4])`, but this is such a common function that there's no reason why we shouldn't allow it.
Also there will be the lazy `substitute` soon: https://github.com/dlang/phobos/pull/5596

File: std/array.d (hunk 2 above)
Comment (wilzbach) [reply std/array.d]: I split it into two lines.

**Comment by quickfur** [comment]
Nice. Well, let's merge this first, then we can fix 10930 next.

**Pull request merged**